exec/supervisor/plugin.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2019 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supervisor
+
+import (
+	"sync"
+
+	pm "github.com/ligato/cn-infra/exec/processmanager"
+	"github.com/ligato/cn-infra/exec/processmanager/status"
+	"github.com/ligato/cn-infra/infra"
+	"github.com/pkg/errors"
+)
+
+const unnamedProcess = "<unnamed>"
+
+// Supervisor defines methods to obtain information about running instances
+type Supervisor interface {
+	// GetProcessNames returns names of all running process instances
+	GetProcessNames() []string
+	// GetProcessByName returns an instance of given process
+	GetProcessByName(name string) pm.ProcessInstance
+}
+
+// Plugin is a supervisor plugin structure
+type Plugin struct {
+	// a list of active process instances
+	mx        sync.Mutex
+	instances map[string]pm.ProcessInstance
+
+	config *Config
+	Deps
+}
+
+// Deps are supervisor dependencies (standard infra plugins + process manager)
+type Deps struct {
+	pm pm.ProcessManager
+	infra.PluginDeps
+}
+
+// Init starts processes and their watchers defined by the configuration file
+func (p *Plugin) Init() error {
+	p.instances = make(map[string]pm.ProcessInstance)
+
+	if err := p.getConfig(); err != nil {
+		return err
+	}
+
+	// run watchers even if there are failed processes. They will be handled
+	// as terminated and exit immediately
+	err := p.start(p.config.Processes)
```

Comment:
We don't want to start watch if there is an error. Check error before calling watch.

Reply:
The watch is not started (I should rewrite the comment), the process is handled as terminated. Another option is to stop processes which were previously started before (it there were any) and return the error.